tools/moduler/cmd/root.go

```diff
@@ -0,0 +1,238 @@
+			}
+		}
+	}
+	return nil
+}
+
+func readNewTags(root string) ([]string, error) {
+	// get list of all existing tags
+	tags, err := getTagsHook(root)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list tags: %v", err)
+	}
+	printf("Get %d tags from remote\n", len(tags))
+	vprintln(strings.Join(tags, "\n"))
+	// find all version.go files
+	files, err := findAllFiles(root, versionFile)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find all version.go files: %v", err)
+	}
+	// read new tags in version.go
+	newTags, err := readTags(tags, files)
+	if err != nil {
+		return nil, fmt.Errorf("failed during reading new tags: %v", err)
+	}
+	return newTags, nil
+}
+
+func getTags(repoPath string) ([]string, error) {
+	wt, err := repo.Get(repoPath)
+	if err != nil {
+		return nil, err
+	}
+	return wt.ListTags("*")
+}
+
+func tagExists(tags []string, tag string) bool {
+	for _, item := range tags {
+		if item == tag {
+			return true
+		}
+	}
+	return false
+}
+
+func findAllFiles(root, filename string) ([]string, error) {
+	// check if root exists
+	if _, err := os.Stat(root); os.IsNotExist(err) {
+		return nil, fmt.Errorf("the root path '%s' does not exist", root)
+	}
+	files := make([]string, 0)
+	err := filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
+		if err == nil && !info.IsDir() && info.Name() == filename {
+			files = append(files, path)
+			return nil
+		}
+		return nil
+	})
+	return files, err
+}
+
+func readTags(tags []string, files []string) ([]string, error) {
+	var newTags []string
+	for _, file := range files {
+		newTag, err := readNewTagInFile(file)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read tag in file '%s': %v", file, err)
+		}
+		if newTag == "" {
+			printf("Found empty tag in file '%s'\n", file)
+			continue
+		} else if tagExists(tags, newTag) {
+			vprintf("Found existed tag '%s' in file '%s'\n", newTag, file)
+			continue
+		}
+		// add new tag to list
+		tags = append(tags, newTag)
+		newTags = append(newTags, newTag)
+	}
+	return newTags, nil
+}
+
+func readNewTagInFile(path string) (string, error) {
+	file, err := os.OpenFile(path, os.O_RDWR, 0666)
+	defer file.Close()
+	if err != nil {
+		return "", err
+	}
+	scanner := bufio.NewScanner(file)
+	scanner.Split(bufio.ScanLines)
+	tag := ""
+	for scanner.Scan() {
+		line := scanner.Text()
+		if strings.HasPrefix(line, tagPrefix) {
+			tag = strings.ReplaceAll(line, tagPrefix, "")
+			break
+		}
+	}
+	return tag, nil
+}
+
+func addNewTag(newTag string) error {
+	vprintf("Adding new tag '%s'\n", newTag)
+	if err := executeCommand("git", "tag", "-a", newTag, "-m", newTag); err != nil {
+		return err
+	}
+	return nil
+}
+
+func pushNewTags(newTags []string) error {
+	vprintln("Pushing new tags")
+	for _, tag := range newTags {
+		if err := addNewTag(tag); err != nil {
+			return err
+		}
+	}
+	return executeCommand("git", "push", "origin", "--tags")
+}
+
+func generateProfiles(profilePath string) error {
```

Comment:
This should happen as part of the build pipeline.  Each tool should be responsible for one job, then they can be composed via scripts etc.  See also https://en.wikipedia.org/wiki/Unix_philosophy

Reply:
Sure, I will move this back to after_script.